android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/VPNTileService.kt

```diff
@@ -0,0 +1,137 @@
+                mService = null
+            } catch (e: RemoteException) {
+                e.printStackTrace()
+            }
+        }
+        override fun onServiceDisconnected(p0: ComponentName?) {
+            mState = State.Error
+            updateTile()
+        }
+    }
+
+    override fun onStartListening() {
+        super.onStartListening()
+        bindService(
+            Intent(this, VPNService::class.java), mServiceConnection, BIND_AUTO_CREATE
+        )
+    }
+
+    override fun onStopListening() {
+        super.onStopListening()
+        unbindService(mServiceConnection)
+    }
+
+    override fun onClick() {
+        when (mState) {
+            State.Error -> return
+            State.Unknown -> return
+            State.Connected -> {
+                if (isLocked) {
+                    // We Should not allow the vpn to be disabled
+                    // while the screen is locked
+                    return
+                }
+                mService?.transact(VPNServiceBinder.ACTIONS.deactivate, Parcel.obtain(), Parcel.obtain(), 0)
+                mState = State.Disconnected
+            }
+            State.Disconnected -> {
+                // No Need to check for secure :)
+                mService?.transact(VPNServiceBinder.ACTIONS.reactivate, Parcel.obtain(), Parcel.obtain(), 0)
+                mState = State.Connected
+            }
+        }
+        updateTile()
+        super.onClick()
+    }
+
+    private fun updateTile() {
+        qsTile.apply {
+            when (mState) {
+                State.Error -> {
+                    state = Tile.STATE_UNAVAILABLE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Error CommuniCating with service"
+                    }
+                }
+                State.Unknown -> {
+                    state = Tile.STATE_UNAVAILABLE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Not Connected With Service"
+                    }
+                }
+                State.Connected -> {
+                    state = Tile.STATE_ACTIVE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Connected with VPN"
+                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                            subtitle = mCity
+                        }
+                    }
+                }
+                State.Disconnected -> {
+                    state = Tile.STATE_INACTIVE
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                        subtitle = "Disconnected"
+                    }
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                        stateDescription = "Not Connected with VPN"
+                    }
+                }
+            }
+        }
```

Comment:
That is some annoying version checking... Does it error without it? Android should just ignore the unknown stuff...

Reply:
Removed some redudancy :) 
Not checking not an option as ```stateDescription = "Not Connected with VPN"``` will be translated into ```qsTile.setDescription(<>)``` and calling nonexistent functions is nothing the jvm likes >:c

Reply:
Boo JVM!